### PR TITLE
Parse meta element in report.xml and use data for generating test cases

### PR DESCRIPTION
**Pull request by fonglh**
The generated JUnit XML report from running the evaluator can return a meta element with some additional information about the test case. These are the `expression`, `expected value` and the `hint`. 

This PR does the following:

1. Add support for additional data in the test case model.
2. Parse the XML for the meta data.
3. Use the meta data when generating the test case.
4. Remove references to `description` and `public` from the views.

Showing the extra metadata will be done in a separate PR.

Tests for inferring the test case type can be improved.

References #1198 and #1200.

### CODE REVIEW

File: app/models/course/assessment/question/programming_test_case.rb [hunk 1]
```diff
@@ -1,9 +1,7 @@
 # frozen_string_literal: true
 class Course::Assessment::Question::ProgrammingTestCase < ActiveRecord::Base
-  schema_validations except: :description
-
-  validates :description, exclusion: [nil]
-
   belongs_to :question, class_name: Course::Assessment::Question::Programming.name,
                         inverse_of: :test_cases
+
+  enum test_case_type: { private_test: 0, public_test: 1, evaluation_test: 2 }
```
Comment (allenwq) [inline app/models/course/assessment/question/programming_test_case.rb]: Move this before associations, check the attributes order in: https://github.com/Coursemology/coursemology2/blob/master/CONTRIBUTING.md

File: lib/autoload/course/assessment/programming_test_case_report.rb [hunk 2]
```diff
@@ -94,6 +94,36 @@ def identifier
       "#{@test_suite.identifier}/#{class_name}#{name.underscore}"
     end
 
+    # The test expression
+    #
+    # @return [String]
+    def expression
+      @expression ||= begin
+        expression = @test_case.search('meta')
+        expression.present? ? expression[0]['expression'] : ''
```
Comment (allenwq) [inline lib/autoload/course/assessment/programming_test_case_report.rb]: actually should this be `meta[0]` instead ?

File: lib/autoload/course/assessment/programming_test_case_report.rb [hunk 3]
```diff
@@ -94,6 +94,36 @@ def identifier
...
+      end
+    end
+
+    # The expected value from running the test expression
+    #
+    # @return [String]
+    def expected
+      @expected ||= begin
+        expected = @test_case.search('meta')
```
Comment (allenwq) [inline lib/autoload/course/assessment/programming_test_case_report.rb]: I think these can be factored out to a meta method.

File: spec/libraries/course/assessment/programming_test_case_report_spec.rb [hunk 4]
```diff
@@ -144,4 +144,59 @@
       end
     end
   end
+
+  context 'when given a report with test case meta information' do
+    self::REPORT_PATH = File.join(Rails.root,
```
Comment (allenwq) [inline spec/libraries/course/assessment/programming_test_case_report_spec.rb]: let(:report_file) will do, don't need to use a constant.

File: spec/libraries/course/assessment/programming_test_case_report_spec.rb [hunk 5]
```diff
@@ -144,4 +144,59 @@
...
+                                  'spec/fixtures/course/'\
+                                  'programming_single_test_suite_report_meta.xml')
+    self::REPORT_XML = File.read(self::REPORT_PATH)
```
Comment (allenwq) [inline spec/libraries/course/assessment/programming_test_case_report_spec.rb]: Same here.

File: lib/autoload/course/assessment/programming_test_case_report.rb (hunk 2 above)
Comment (fonglh) [inline lib/autoload/course/assessment/programming_test_case_report.rb]: Yes that'll be better, I'll rename the variables.

File: app/services/course/assessment/question/programming_import_service.rb [hunk 6]
```diff
@@ -103,8 +103,24 @@ def build_test_case_records(test_report)
...
+                                 hint: test_case.hint)
+    end
+  end
+
+  # Figures out what kind of test case it is from the name
+  #
+  # @param [String] test_case_name The name of the test case.
+  # @return [Symbol]
+  def infer_test_case_type(test_case_name)
+    if test_case_name =~ /public/i
```
Comment (lowjoel) [inline app/services/course/assessment/question/programming_import_service.rb]: I vaguely remember that a case-when with regexes will properly match the string.

**Pull request merged**